active/003-seq-feature.md

```diff
@@ -0,0 +1,98 @@
+
+Additional benefits of having a `interval_metadata` objects include the ability to rapidly query features by interval.  IntervalTrees allow for both querying by position and ranges to retrieve all overlapping intervals.  This can be particularly helpful when dealing with layers of features (i.e. transcripts made up of exons, or operons made up of genes).
+
+# Detailed design
+
+We propose to use the `IntervalTree` object in `bx-python.`  Some benchmarks have been made for various IntervalTree data structures - in this [benchmark](https://gist.github.com/shoyer/c939325f509d7c027949), bx-python was determined to have one of the fastest query times.
+
+## `IntervalMetadata` object structure
+features : dictionary.  keys are `skbio.Feature` objects.  values are a list of tuples, where each tuple corresponds to an interval.
+intervals : bx-python `IntervalTree` object.  The keys correspond to intervals and the values correspond to a single `skbio.Feature` object.   
+
+These objects are arranged such that features can be queried by both feature attributes and intervals.
+
+## `Feature` object structure
+The feature object is more a less a immutable, hashable dictionary.  This store arbiturary information about features.  For instance, strand information, CDS, ...
+Since there is no standard in genbank and related formats for keywords, arbituary keywords can be encoded
+```python
+gene = Feature(name='sagA', type='CDS', strand='+', function='toxin')
+```
+The above code would encode for a gene, whose name is 'sagA', that is a coding region on the positive strand whose function is a toxin.
+
+## `IntervalMetadata` functions
+`constructor(features=None)`
+- `features` : dictionary.  keys are `skbio.Feature` objects.  values are a list of tuples, where each tuple corresponds to an interval.
+- Called to initialize the `IntervalMetadata` object
+
+`reverse_complement(length)`
+- `length` : int.  The length of the genome.  This is required when swapping coordinates
+- This performs a reverse complement on all the coordinates with in IntervalTree.
+
+```python
+   interval_metadata = IntervalMetadata()
+   interval_metadata.add(Feature(gene='sagB', location=0), (3, 5))
+   iv = interval_metadata.reverse_complement(length=10)
+```
+
+`update(old_feature, new_feature)`
+- `old_feature` : `skbio.Feature`.  feature to be replaced
+- `new_feature` : `skbio.Feature`.  updated feature to replace the original feature
+- Since `skbio.Feature` objects are immutable, if individual features in the `IntervalMetadata` are to be updated, the need to be completely replaced.
+
+```python
+    interval_metadata = IntervalMetadata()
+        interval_metadata.add(Feature(gene='sagA', location=0), 1, (4, 7))
+        interval_metadata.add(Feature(gene='sagB', location=0), (3, 5))
+        interval_metadata.update(Feature(gene='sagB', location=0),
+                                 Feature(gene='sagB', location=1))
+```
+
+
+`add(feature, *intervals)`
```

Comment:
@ebolyen, in your example, would that mean that genome positions 1, 4-7 (exclusive), 8, and 10-15 (exclusive) would assigned a specific feature? The reason I ask is that I think we need to be careful about the overloading here it is easy to look at this in the context of `np.array` where this would be indexing into a 4-dimensional matrix.